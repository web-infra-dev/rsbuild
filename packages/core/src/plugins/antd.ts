import { getAntdMajorVersion } from '@modern-js/utils';
import type { RsbuildTarget, DefaultRsbuildPlugin } from '@rsbuild/shared';

export const pluginAntd = (): DefaultRsbuildPlugin => ({
  name: `plugin-antd`,

  setup(api) {
    api.modifyRsbuildConfig((builderConfig) => {
      builderConfig.source ??= {};

      if (
        builderConfig.source.transformImport === false ||
        builderConfig.source.transformImport?.some(
          (item) => item.libraryName === 'antd',
        )
      ) {
        return;
      }

      const antdMajorVersion = getAntdMajorVersion(api.context.rootPath);
      // antd >= v5 no longer need babel-plugin-import
      // see: https://ant.design/docs/react/migration-v5#remove-babel-plugin-import
      if (antdMajorVersion && antdMajorVersion < 5) {
        builderConfig.source ??= {};
        builderConfig.source.transformImport = [
          ...(builderConfig.source.transformImport || []),
          {
            libraryName: 'antd',
            libraryDirectory: useSSR(api.context.target) ? 'lib' : 'es',
            style: true,
          },
        ];
      }
    });
  },
});

export function useSSR(target: RsbuildTarget | RsbuildTarget[]) {
  return (Array.isArray(target) ? target : [target]).some((item) =>
    ['node', 'service-worker'].includes(item),
  );
}
