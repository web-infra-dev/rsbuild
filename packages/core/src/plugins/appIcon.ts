import path from 'node:path';
import { promisify } from 'node:util';
import { lookup } from 'mrmime';
import {
  ensureAssetPrefix,
  fileExistsByCompilation,
  getPublicPathFromCompiler,
  isURL,
} from '../helpers';
import type { AppIconItem, HtmlBasicTag, RsbuildPlugin } from '../types';

type IconExtra = {
  sizes: string;
  isURL?: boolean;
  absolutePath: string;
  relativePath: string;
  requestPath: string;
  mimeType?: string;
};

export const pluginAppIcon = (): RsbuildPlugin => ({
  name: 'rsbuild:app-icon',

  setup(api) {
    const htmlTagsMap = new Map<string, HtmlBasicTag[]>();
    const iconFormatMap = new Map<string, IconExtra>();

    const formatIcon = (
      icon: AppIconItem,
      distDir: string,
      publicPath: string,
    ): AppIconItem & IconExtra => {
      const { src, size } = icon;
      const cached = iconFormatMap.get(src);

      if (cached) {
        return { ...cached, ...icon };
      }

      const sizes = `${size}x${size}`;

      if (isURL(src)) {
        const paths = {
          sizes,
          isURL: true,
          requestPath: src,
          absolutePath: src,
          relativePath: src,
          mimeType: lookup(src),
        };

        iconFormatMap.set(src, paths);

        return { ...paths, ...icon };
      }

      const absolutePath = path.isAbsolute(src)
        ? src
        : path.join(api.context.rootPath, src);
      const relativePath = path.posix.join(
        distDir,
        path.basename(absolutePath),
      );
      const requestPath = ensureAssetPrefix(relativePath, publicPath);

      const paths = {
        sizes,
        requestPath,
        absolutePath,
        relativePath,
        mimeType: lookup(absolutePath),
      };

      iconFormatMap.set(src, paths);

      return { ...paths, ...icon };
    };

    api.processAssets(
      { stage: 'additional' },
      async ({ compilation, environment, sources }) => {
        const { config } = environment;
        const { appIcon } = config.html;

        if (!appIcon) {
          return;
        }

        const distDir = config.output.distPath.image;
        const manifestFile = appIcon.filename ?? 'manifest.webmanifest';
        const publicPath = getPublicPathFromCompiler(compilation);
        const icons = appIcon.icons.map((icon) =>
          formatIcon(icon, distDir, publicPath),
        );
        const tags: HtmlBasicTag[] = [];

        for (const icon of icons) {
          if (icon.target === 'web-app-manifest' && !appIcon.name) {
            throw new Error(
              "[rsbuild:app-icon] `appIcon.name` is required when `target` is 'web-app-manifest'.",
            );
          }

          if (!icon.isURL) {
            if (!compilation.inputFileSystem) {
              throw new Error(
                `[rsbuild:app-icon] 'compilation.inputFileSystem' is not available.`,
              );
            }

            if (
              !(await fileExistsByCompilation(compilation, icon.absolutePath))
            ) {
              throw new Error(
                `[rsbuild:app-icon] Can not find the app icon, please check if the '${icon.relativePath}' file exists'.`,
              );
            }

            const source = await promisify(
              compilation.inputFileSystem.readFile,
            )(icon.absolutePath);

            if (!source) {
              throw new Error(
                `[rsbuild:app-icon] Failed to read the app icon file, please check if the '${icon.relativePath}' file exists'.`,
              );
            }

            compilation.emitAsset(
              icon.relativePath,
              new sources.RawSource(source),
            );
          }

          if (
            icon.target === 'apple-touch-icon' ||
            (!icon.target && icon.size < 200)
          ) {
            tags.push({
              tag: 'link',
              attrs: {
                rel: 'apple-touch-icon',
                sizes: icon.sizes,
                href: icon.requestPath,
              },
            });
          }
        }

        if (appIcon.name) {
          const manifestIcons = icons
            .filter(
              (icon) => icon.target === 'web-app-manifest' || !icon.target,
            )
            .map((icon) => {
              const result = {
                src: icon.requestPath,
                sizes: icon.sizes,
              };
              if (icon.mimeType) {
                return { ...result, type: icon.mimeType };
              }
              return result;
            });

          const manifest = {
            name: appIcon.name,
            icons: manifestIcons,
          };

          compilation.emitAsset(
            manifestFile,
            new sources.RawSource(JSON.stringify(manifest)),
          );

          tags.push({
            tag: 'link',
            attrs: {
              rel: 'manifest',
              href: ensureAssetPrefix(manifestFile, publicPath),
            },
          });
        }

        if (tags.length) {
          htmlTagsMap.set(environment.name, tags);
        }
      },
    );

    api.modifyHTMLTags(({ headTags, bodyTags }, { environment }) => {
      const tags = htmlTagsMap.get(environment.name);
      if (tags) {
        headTags.unshift(...tags);
      }
      return { headTags, bodyTags };
    });

    api.onCloseDevServer(() => {
      htmlTagsMap.clear();
      iconFormatMap.clear();
    });
  },
});
