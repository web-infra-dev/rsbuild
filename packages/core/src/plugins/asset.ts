import path from 'path';
import {
  getDistPath,
  getFilename,
  chainStaticAssetRule,
} from '@rsbuild/shared';
import type { DefaultRsbuildPlugin } from '@rsbuild/shared';

export function getRegExpForExts(exts: string[]): RegExp {
  const matcher = exts
    .map((ext) => ext.trim())
    .map((ext) => (ext.startsWith('.') ? ext.slice(1) : ext))
    .join('|');

  return new RegExp(
    exts.length === 1 ? `\\.${matcher}$` : `\\.(${matcher})$`,
    'i',
  );
}

export const pluginAsset = (
  assetType: 'image' | 'media' | 'font' | 'svg',
  exts: string[],
): DefaultRsbuildPlugin => ({
  name: `plugin-${assetType}`,

  setup(api) {
    api.modifyBundlerChain((chain, { isProd }) => {
      const config = api.getNormalizedConfig();
      const regExp = getRegExpForExts(exts);
      const distDir = getDistPath(config.output, assetType);
      const filename = getFilename(config.output, assetType, isProd);

      const maxSize = config.output.dataUriLimit[assetType];

      const rule = chain.module.rule(assetType).test(regExp);

      chainStaticAssetRule({
        rule,
        maxSize,
        filename: path.posix.join(distDir, filename),
        assetType,
      });
    });
  },
});
