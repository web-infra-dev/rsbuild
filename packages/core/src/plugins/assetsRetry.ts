import {
  getDistPath,
  isHtmlDisabled,
  DefaultRsbuildPlugin,
} from '@rsbuild/shared';

export function pluginAssetsRetry(): DefaultRsbuildPlugin {
  return {
    name: 'plugin-assets-retry',
    setup(api) {
      api.modifyBundlerChain(
        async (chain, { CHAIN_ID, target, HtmlPlugin }) => {
          const config = api.getNormalizedConfig();

          if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
            return;
          }

          const { AssetsRetryPlugin } = await import('@rsbuild/shared');
          const distDir = getDistPath(config.output, 'js');

          const { assetsRetry = {} } = config.output;

          // assetsRetry.crossOrigin should be same as html.crossorigin by default
          if (assetsRetry.crossOrigin === undefined) {
            assetsRetry.crossOrigin = config.html.crossorigin;
          }

          chain.plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY).use(AssetsRetryPlugin, [
            {
              ...assetsRetry,
              distDir,
              HtmlPlugin,
            },
          ]);
        },
      );
    },
  };
}
