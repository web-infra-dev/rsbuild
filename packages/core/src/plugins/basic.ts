import path from 'node:path';
import { TARGET_ID_MAP, getJsSourceMap, isUsingHMR } from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

/**
 * Provide some basic configs of rspack
 */
export const pluginBasic = (): RsbuildPlugin => ({
  name: 'rsbuild:basic',

  setup(api) {
    api.modifyBundlerChain(
      (chain, { env, isProd, target, bundler, CHAIN_ID }) => {
        const config = api.getNormalizedConfig();

        chain.name(TARGET_ID_MAP[target]);

        chain.devtool(getJsSourceMap(config));

        // The base directory for resolving entry points and loaders from the configuration.
        chain.context(api.context.rootPath);

        chain.mode(isProd ? 'production' : 'development');

        chain.merge({
          infrastructureLogging: {
            // Using `error` level to avoid `cache.PackFileCacheStrategy` logs
            level: 'error',
          },
        });

        // Disable performance hints, these logs are too complex
        chain.performance.hints(false);

        // Align with the futureDefaults of webpack 6
        chain.module.parser.merge({
          javascript: {
            exportsPresence: 'error',
          },
        });

        const usingHMR = isUsingHMR(config, { target, isProd });

        if (usingHMR) {
          chain
            .plugin(CHAIN_ID.PLUGIN.HMR)
            .use(bundler.HotModuleReplacementPlugin);
        }

        if (env === 'development') {
          // Set correct path for source map
          // this helps VS Code break points working correctly in monorepo
          chain.output.devtoolModuleFilenameTemplate(
            (info: { absoluteResourcePath: string }) =>
              path.resolve(info.absoluteResourcePath).replace(/\\/g, '/'),
          );
        }
      },
    );
  },
});
