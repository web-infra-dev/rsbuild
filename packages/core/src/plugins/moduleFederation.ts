import type { CacheGroup, Rspack } from '@rsbuild/shared';
import { rspack } from '@rspack/core';
import type { RspackPluginInstance } from '@rspack/core';
import { DEFAULT_ASSET_PREFIX } from '../constants';
import type { RsbuildPlugin } from '../types';

/**
 * Force remote entry not be affected by user's chunkSplit strategy,
 * Otherwise, the remote chunk will not be loaded correctly.
 * @see https://github.com/web-infra-dev/rsbuild/discussions/1461#discussioncomment-8329790
 */
class PatchSplitChunksPlugin implements RspackPluginInstance {
  name: string;

  constructor(name: string) {
    this.name = name;
  }

  apply(compiler: Rspack.Compiler): void {
    const { splitChunks } = compiler.options.optimization;

    if (!splitChunks) {
      return;
    }

    const applyPatch = (cacheGroup: CacheGroup) => {
      if (typeof cacheGroup !== 'object' || cacheGroup instanceof RegExp) {
        return;
      }

      // cacheGroup.chunks will inherit splitChunks.chunks
      // so we only need to modify the chunks that are set separately.
      const { chunks } = cacheGroup;
      if (!chunks) {
        return;
      }

      if (typeof chunks === 'function') {
        const prevChunks = chunks;

        cacheGroup.chunks = (chunk) => {
          if (chunk.name && chunk.name === this.name) {
            return false;
          }
          return prevChunks(chunk);
        };
        return;
      }

      if (chunks === 'all') {
        cacheGroup.chunks = (chunk) => {
          if (chunk.name && chunk.name === this.name) {
            return false;
          }
          return true;
        };
        return;
      }

      if (chunks === 'initial') {
        cacheGroup.chunks = (chunk) => {
          if (chunk.name && chunk.name === this.name) {
            return false;
          }
          return chunk.isOnlyInitial();
        };
        return;
      }
    };

    // patch splitChunk.chunks
    applyPatch(splitChunks);

    const { cacheGroups } = splitChunks;
    if (!cacheGroups) {
      return;
    }

    // patch splitChunk.cacheGroups[key].chunks
    for (const cacheGroupKey of Object.keys(cacheGroups)) {
      applyPatch(cacheGroups[cacheGroupKey]);
    }
  }
}

export function pluginModuleFederation(): RsbuildPlugin {
  return {
    name: 'rsbuild:module-federation',

    setup(api) {
      api.modifyEnvironmentConfig((config) => {
        /**
         * Currently, splitChunks will take precedence over module federation shared modules.
         * So we need to disable the default split chunks rules to make shared modules to work properly.
         * @see https://github.com/module-federation/module-federation-examples/issues/3161
         */
        if (
          config.moduleFederation?.options &&
          config.performance?.chunkSplit?.strategy === 'split-by-experience'
        ) {
          config.performance.chunkSplit = {
            ...config.performance.chunkSplit,
            strategy: 'custom',
          };
        }
      });

      api.modifyBundlerChain(
        async (chain, { CHAIN_ID, target, environment }) => {
          const config = api.getNormalizedConfig({ environment });

          if (!config.moduleFederation?.options || target !== 'web') {
            return;
          }

          const { options } = config.moduleFederation;

          chain
            .plugin(CHAIN_ID.PLUGIN.MODULE_FEDERATION)
            .use(rspack.container.ModuleFederationPlugin, [options]);

          if (options.name) {
            chain
              .plugin('mf-patch-split-chunks')
              .use(PatchSplitChunksPlugin, [options.name]);
          }

          const publicPath = chain.output.get('publicPath');

          // set the default publicPath to 'auto' to make MF work
          if (publicPath === DEFAULT_ASSET_PREFIX) {
            chain.output.set('publicPath', 'auto');
          }
        },
      );
    },
  };
}
