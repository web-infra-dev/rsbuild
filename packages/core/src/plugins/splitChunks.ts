import assert from 'node:assert';
import {
  NODE_MODULES_REGEX,
  createDependenciesRegExp,
  type Polyfill,
  type CacheGroups,
  type SplitChunks,
  type ForceSplitting,
  type RsbuildChunkSplit,
} from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

// We expose the three-layer to specify webpack chunk-split ability:
// 1. By strategy.There some best practice integrated in our internal strategy.
// 2. By forceSplitting config, which is designed to split chunks by user defined rules.That's easier to use than webpack raw config.
// 3. By override config, which is designed to override the raw config of webpack `splitChunks`.It has the highest priority.
// By the way, the config complexity is increasing gradually.

interface SplitChunksContext {
  /**
   * User defined cache groups which can be reused across different split strategies
   */
  userDefinedCacheGroups: CacheGroups;
  /**
   * Default split config in webpack
   */
  defaultConfig: Exclude<SplitChunks, false>;
  /**
   * User webpack `splitChunks` config
   */
  override: Exclude<SplitChunks, false>;
  /**
   * User Rsbuild `chunkSplit` config
   */
  userConfig: RsbuildChunkSplit;
  /**
   * The root path of current project
   */
  rootPath: string;
  /**
   * The polyfill mode.
   */
  polyfill: Polyfill;
}

function getUserDefinedCacheGroups(
  forceSplitting: ForceSplitting,
): CacheGroups {
  const cacheGroups: CacheGroups = {};
  const pairs = Array.isArray(forceSplitting)
    ? forceSplitting.map(
        (regexp, index) => [`force-split-${index}`, regexp] as const,
      )
    : Object.entries(forceSplitting);

  for (const [key, regexp] of pairs) {
    cacheGroups[key] = {
      test: regexp,
      name: key,
      chunks: 'all',
      // Ignore minimum size, minimum chunks and maximum requests and always create chunks for user defined cache group.
      enforce: true,
    };
  }

  return cacheGroups;
}

function splitByExperience(ctx: SplitChunksContext): SplitChunks {
  const { override, polyfill, defaultConfig, userDefinedCacheGroups } = ctx;
  const experienceCacheGroup: CacheGroups = {};

  const packageRegExps: Record<string, RegExp> = {
    lodash: createDependenciesRegExp('lodash', 'lodash-es'),
    axios: createDependenciesRegExp('axios', /axios-.+/),
  };

  if (polyfill === 'entry' || polyfill === 'usage') {
    packageRegExps.polyfill = createDependenciesRegExp(
      'tslib',
      'core-js',
      '@babel/runtime',
      '@swc/helpers',
    );
  }

  for (const [name, test] of Object.entries(packageRegExps)) {
    const key = `lib-${name}`;

    experienceCacheGroup[key] = {
      test,
      priority: 0,
      name: key,
      reuseExistingChunk: true,
    };
  }

  return {
    ...defaultConfig,
    ...override,
    cacheGroups: {
      ...defaultConfig.cacheGroups,
      ...experienceCacheGroup,
      ...userDefinedCacheGroups,
      ...override.cacheGroups,
    },
  };
}

export const MODULE_PATH_REGEX =
  /.*[\\/]node_modules[\\/](?!\.pnpm[\\/])(?:(@[^\\/]+)[\\/])?([^\\/]+)/;

export function getPackageNameFromModulePath(modulePath: string) {
  const handleModuleContext = modulePath?.match(MODULE_PATH_REGEX);

  if (!handleModuleContext) {
    return undefined;
  }

  const [, scope, name] = handleModuleContext;
  const packageName = ['npm', (scope ?? '').replace('@', ''), name]
    .filter(Boolean)
    .join('.');

  return packageName;
}

function splitByModule(ctx: SplitChunksContext): SplitChunks {
  const { override, userDefinedCacheGroups, defaultConfig } = ctx;
  return {
    ...defaultConfig,
    minSize: 0,
    maxInitialRequests: Number.POSITIVE_INFINITY,
    ...override,
    cacheGroups: {
      ...defaultConfig.cacheGroups,
      ...userDefinedCacheGroups,
      // Core group
      vendors: {
        priority: -10,
        test: NODE_MODULES_REGEX,
        name(module) {
          return module
            ? getPackageNameFromModulePath(module.context!)
            : undefined;
        },
      },
      ...override.cacheGroups,
    },
  };
}

function splitBySize(ctx: SplitChunksContext): SplitChunks {
  const { override, userDefinedCacheGroups, defaultConfig, userConfig } = ctx;
  assert(userConfig.strategy === 'split-by-size');

  return {
    ...defaultConfig,
    minSize: userConfig.minSize ?? 0,
    maxSize: userConfig.maxSize ?? Number.POSITIVE_INFINITY,
    ...override,
    cacheGroups: {
      ...defaultConfig.cacheGroups,
      ...userDefinedCacheGroups,
      ...override.cacheGroups,
    },
  };
}

function splitCustom(ctx: SplitChunksContext): SplitChunks {
  const { override, userDefinedCacheGroups, defaultConfig } = ctx;
  return {
    ...defaultConfig,
    ...override,
    cacheGroups: {
      ...defaultConfig.cacheGroups,
      ...userDefinedCacheGroups,
      ...override.cacheGroups,
    },
  };
}

function allInOne(_ctx: SplitChunksContext): SplitChunks {
  // Set false to avoid chunk split.
  return false;
}

// Ignore user defined cache group to get single vendor chunk.
function singleVendor(ctx: SplitChunksContext): SplitChunks {
  const { override, defaultConfig, userDefinedCacheGroups } = ctx;

  const singleVendorCacheGroup: CacheGroups = {
    singleVendor: {
      test: NODE_MODULES_REGEX,
      priority: 0,
      chunks: 'all',
      name: 'vendor',
      enforce: true,
      reuseExistingChunk: true,
    },
  };

  return {
    ...defaultConfig,
    ...override,
    cacheGroups: {
      ...defaultConfig.cacheGroups,
      ...singleVendorCacheGroup,
      ...userDefinedCacheGroups,
      ...override.cacheGroups,
    },
  };
}

const SPLIT_STRATEGY_DISPATCHER: Record<
  string,
  (ctx: SplitChunksContext) => SplitChunks | Promise<SplitChunks>
> = {
  'split-by-experience': splitByExperience,
  'split-by-module': splitByModule,
  'split-by-size': splitBySize,
  custom: splitCustom,
  'all-in-one': allInOne,
  'single-vendor': singleVendor,
};

export const pluginSplitChunks = (): RsbuildPlugin => ({
  name: 'rsbuild:split-chunks',
  setup(api) {
    api.modifyBundlerChain(
      async (chain, { isServer, isWebWorker, isServiceWorker }) => {
        if (isServer || isWebWorker || isServiceWorker) {
          chain.optimization.splitChunks(false);

          // web worker does not support dynamic imports, dynamicImportMode need set to eager
          if (isWebWorker || isServiceWorker) {
            chain.module.parser.merge({
              javascript: {
                dynamicImportMode: 'eager',
              },
            });
          }

          return;
        }

        const config = api.getNormalizedConfig();
        const defaultConfig: Exclude<SplitChunks, false> = {
          // Optimize both `initial` and `async` chunks
          chunks: 'all',
          // When chunk size >= 50000 bytes, split it into separate chunk
          // @ts-expect-error Rspack type missing
          enforceSizeThreshold: 50000,
          cacheGroups: {},
        };

        const { chunkSplit } = config.performance;
        let userDefinedCacheGroups = {};

        if (chunkSplit.forceSplitting) {
          userDefinedCacheGroups = getUserDefinedCacheGroups(
            chunkSplit.forceSplitting,
          );
        }

        // Patch the override config difference between the `custom` strategy and other strategy.
        const override =
          chunkSplit.strategy === 'custom'
            ? chunkSplit.splitChunks ?? chunkSplit.override
            : chunkSplit.override;

        // Apply different strategy
        const splitChunksOptions = await SPLIT_STRATEGY_DISPATCHER[
          chunkSplit.strategy || 'split-by-experience'
        ]({
          defaultConfig,
          override: override || {},
          userDefinedCacheGroups,
          userConfig: chunkSplit,
          rootPath: api.context.rootPath,
          polyfill: config.output.polyfill,
        });

        chain.optimization.splitChunks(
          // @ts-expect-error splitChunks type mismatch
          splitChunksOptions,
        );
      },
    );
  },
});
