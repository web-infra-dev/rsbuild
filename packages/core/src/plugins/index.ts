import {
  MEDIA_EXTENSIONS,
  FONT_EXTENSIONS,
  IMAGE_EXTENSIONS,
  Plugins,
} from '@rsbuild/shared';

export const plugins: Plugins = {
  html: () => import('./html').then((m) => m.pluginHtml()),
  cleanOutput: () => import('./cleanOutput').then((m) => m.pluginCleanOutput()),
  startUrl: () => import('./startUrl').then((m) => m.pluginStartUrl()),
  fileSize: () => import('./fileSize').then((m) => m.pluginFileSize()),
  devtool: () => import('./devtool').then((m) => m.pluginDevtool()),
  target: () => import('./target').then((m) => m.pluginTarget()),
  entry: () => import('./entry').then((m) => m.pluginEntry()),
  cache: () => import('./cache').then((m) => m.pluginCache()),
  yaml: () => import('./yaml').then((m) => m.pluginYaml()),
  toml: () => import('./toml').then((m) => m.pluginToml()),
  svg: () => import('./svg').then((m) => m.pluginSvg()),
  splitChunks: () => import('./splitChunks').then((m) => m.pluginSplitChunks()),
  inlineChunk: () => import('./inlineChunk').then((m) => m.pluginInlineChunk()),
  bundleAnalyzer: () =>
    import('./bundleAnalyzer').then((m) => m.pluginBundleAnalyzer()),
  font: () =>
    import('./asset').then((m) => m.pluginAsset('font', FONT_EXTENSIONS)),
  image: () =>
    import('./asset').then((m) => m.pluginAsset('image', IMAGE_EXTENSIONS)),
  media: () =>
    import('./asset').then((m) => m.pluginAsset('media', MEDIA_EXTENSIONS)),
  assetsRetry: () => import('./assetsRetry').then((m) => m.pluginAssetsRetry()),
  antd: () => import('./antd').then((m) => m.pluginAntd()),
  arco: () => import('./arco').then((m) => m.pluginArco()),
  checkSyntax: () => import('./checkSyntax').then((m) => m.pluginCheckSyntax()),
  rem: () => import('./rem').then((m) => m.pluginRem()),
  wasm: () => import('./wasm').then((m) => m.pluginWasm()),
  moment: () => import('./moment').then((m) => m.pluginMoment()),
  nodeAddons: () => import('./nodeAddons').then((m) => m.pluginNodeAddons()),
  externals: () => import('./externals').then((m) => m.pluginExternals()),
  networkPerformance: () =>
    import('./networkPerformance').then((m) => m.pluginNetworkPerformance()),
  preloadOrPrefetch: () =>
    import('./preloadOrPrefetch').then((m) => m.pluginPreloadOrPrefetch()),
  performance: () => import('./performance').then((m) => m.pluginPerformance()),
  define: () => import('./define').then((m) => m.pluginDefine()),
};
