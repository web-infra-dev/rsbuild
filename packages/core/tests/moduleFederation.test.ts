import { createStubRsbuild } from '@scripts/test-helper';
import { pluginModuleFederation } from '../src/plugins/moduleFederation';
import { pluginSplitChunks } from '../src/plugins/splitChunks';

describe('plugin-module-federation', () => {
  it('should set module federation config', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginSplitChunks(), pluginModuleFederation()],
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-experience',
          },
        },
        moduleFederation: {
          options: {
            name: 'remote',
            exposes: {
              './Button': './src/Button',
            },
            filename: 'remoteEntry.js',
            shared: {
              react: {
                singleton: true,
                requiredVersion: '^18.2.0',
              },
              'react-dom': {
                singleton: true,
                requiredVersion: '^18.2.0',
              },
            },
          },
        },
      },
    });

    const config = await rsbuild.unwrapConfig();
    expect(config).toMatchSnapshot();
  });

  it('should set environment module federation config correctly', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginSplitChunks(), pluginModuleFederation()],
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-experience',
          },
        },
        environments: {
          web: {
            moduleFederation: {
              options: {
                name: 'remote',
                exposes: {
                  './Button': './src/Button',
                },
                filename: 'remoteEntry.js',
                shared: {
                  react: {
                    singleton: true,
                    requiredVersion: '^18.2.0',
                  },
                  'react-dom': {
                    singleton: true,
                    requiredVersion: '^18.2.0',
                  },
                },
              },
            },
          },
          web1: {},
        },
      },
    });

    const configs = await rsbuild.initConfigs();
    expect(configs).toMatchSnapshot();
  });

  it('should set module federation and environment chunkSplit config correctly', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginSplitChunks(), pluginModuleFederation()],
      rsbuildConfig: {
        moduleFederation: {
          options: {
            name: 'remote',
            exposes: {
              './Button': './src/Button',
            },
            filename: 'remoteEntry.js',
            shared: {
              react: {
                singleton: true,
                requiredVersion: '^18.2.0',
              },
              'react-dom': {
                singleton: true,
                requiredVersion: '^18.2.0',
              },
            },
          },
        },
        environments: {
          web: {
            performance: {
              chunkSplit: {
                strategy: 'split-by-experience',
              },
            },
          },
          web1: {
            performance: {
              chunkSplit: {
                strategy: 'all-in-one',
              },
            },
          },
        },
      },
    });

    const configs = await rsbuild.initConfigs();
    expect(configs).toMatchSnapshot();
  });
});
