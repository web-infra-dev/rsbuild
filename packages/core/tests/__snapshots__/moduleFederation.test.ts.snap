// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-module-federation > should set environment module federation config correctly 1`] = `
[
  {
    "optimization": {
      "splitChunks": {
        "cacheGroups": {},
        "chunks": "all",
        "enforceSizeThreshold": 50000,
      },
    },
    "plugins": [
      ModuleFederationPlugin {
        "_options": {
          "exposes": {
            "Button": "src/Button",
          },
          "filename": "remoteEntry.js",
          "name": "remote",
          "shared": {
            "react": {
              "requiredVersion": "^18.2.0",
              "singleton": true,
            },
            "react-dom": {
              "requiredVersion": "^18.2.0",
              "singleton": true,
            },
          },
        },
      },
      PatchSplitChunksPlugin {
        "name": "remote",
      },
    ],
  },
  {
    "optimization": {
      "splitChunks": {
        "cacheGroups": {
          "lib-axios": {
            "name": "lib-axios",
            "priority": 0,
            "reuseExistingChunk": true,
            "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
          },
          "lib-lodash": {
            "name": "lib-lodash",
            "priority": 0,
            "reuseExistingChunk": true,
            "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
          },
          "lib-polyfill": {
            "name": "lib-polyfill",
            "priority": 0,
            "reuseExistingChunk": true,
            "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(tslib\\|core-js\\|@babel\\\\/runtime\\|@swc\\\\/helpers\\)\\[\\\\\\\\/\\]/,
          },
        },
        "chunks": "all",
        "enforceSizeThreshold": 50000,
      },
    },
  },
]
`;

exports[`plugin-module-federation > should set module federation and environment chunkSplit config correctly 1`] = `
[
  {
    "optimization": {
      "splitChunks": {
        "cacheGroups": {},
        "chunks": "all",
        "enforceSizeThreshold": 50000,
      },
    },
    "plugins": [
      ModuleFederationPlugin {
        "_options": {
          "exposes": {
            "Button": "src/Button",
          },
          "filename": "remoteEntry.js",
          "name": "remote",
          "shared": {
            "react": {
              "requiredVersion": "^18.2.0",
              "singleton": true,
            },
            "react-dom": {
              "requiredVersion": "^18.2.0",
              "singleton": true,
            },
          },
        },
      },
      PatchSplitChunksPlugin {
        "name": "remote",
      },
    ],
  },
  {
    "optimization": {
      "splitChunks": false,
    },
    "plugins": [
      ModuleFederationPlugin {
        "_options": {
          "exposes": {
            "Button": "src/Button",
          },
          "filename": "remoteEntry.js",
          "name": "remote",
          "shared": {
            "react": {
              "requiredVersion": "^18.2.0",
              "singleton": true,
            },
            "react-dom": {
              "requiredVersion": "^18.2.0",
              "singleton": true,
            },
          },
        },
      },
      PatchSplitChunksPlugin {
        "name": "remote",
      },
    ],
  },
]
`;

exports[`plugin-module-federation > should set module federation config 1`] = `
{
  "optimization": {
    "splitChunks": {
      "cacheGroups": {},
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
  "plugins": [
    ModuleFederationPlugin {
      "_options": {
        "exposes": {
          "Button": "src/Button",
        },
        "filename": "remoteEntry.js",
        "name": "remote",
        "shared": {
          "react": {
            "requiredVersion": "^18.2.0",
            "singleton": true,
          },
          "react-dom": {
            "requiredVersion": "^18.2.0",
            "singleton": true,
          },
        },
      },
    },
    PatchSplitChunksPlugin {
      "name": "remote",
    },
  ],
}
`;
